py3status/modules/weather_owm.py

```diff
@@ -0,0 +1,755 @@
+IP_LOC = '//data/location'
+IP_GMT_OFF = '//data/datetime/offset_gmt'
+OWM_CLOUD_COVER = '//clouds/all'
+OWM_DESC = '//weather:0/main'
+OWM_DESC_LONG = '//weather:0/description'
+OWM_HUMIDITY = '//main/humidity'
+OWM_PRESSURE = '//main'
+OWM_RAIN = '//rain/3h'
+OWM_SNOW = '//snow/3h'
+OWM_SUNRISE = '//sys/sunrise'
+OWM_SUNSET = '//sys/sunset'
+OWM_TEMP = '//main'
+OWM_WEATHER_ICON = '//weather:0/id'
+OWM_WIND = '//wind'
+
+# Units constants
+RAIN_UNITS = set(['cm', 'ft', 'in', 'mm', 'm', 'yd'])
+SNOW_UNITS = RAIN_UNITS
+TEMP_UNITS = set(['c', 'f', 'k'])
+WIND_UNITS = set(['fsec', 'msec', 'mph', 'kmh'])
+
+# Conversion factors
+FT_FROM_METER = 3.28084
+IN_FROM_MM = 0.0393701
+KMH_FROM_MSEC = 0.277778
+MPH_FROM_MSEC = 2.23694
+
+# Thresholds defaults
+THRESHOLDS = [
+    (-100, '#0FF'),
+    (0, '#00F'),
+    (50, '#0F0'),
+    (150, '#FF0')
+]
+
+
+class OWMException(Exception):
+    pass
+
+
+class Py3status:
+
+    api_key = None
+    cache_timeout = 600
+    forecast_days = 3
+    forecast_include_today = False
+    forecast_text_separator = ' '
+    format = ('{city} {icon} {temperature}{rain}, {description} {forecast}')
+    format_clouds = '{icon} {coverage}%'
+    format_forecast = '{icon}'
+    format_humidity = '{icon} {humidity}%'
+    format_pressure = '{icon} {pressure} hPa'
+    format_rain = '[\?if=amount  {icon} {amount:.0f} {unit}]'
+    format_snow = '[\?if=amount  {icon} {amount:.0f} {unit}]'
+    format_sunrise = '{icon} %-I:%M %p'
+    format_sunset = '{icon} %-I:%M %p'
+    format_temperature = u'{icon} {current:.0f}°{unit}'
+    format_wind = '[\?if=speed  {icon} {speed:.0f} {unit}]'
+    icon_atmosphere = u'🌫'
+    icon_cloud = u'☁'
+    icon_extreme = u'⚠'
+    icon_humidity = u'●'
+    icon_pressure = u'◌'
+    icon_rain = u'🌧'
+    icon_snow = u'❄'
+    icon_sun = u'☼'
+    icon_sunrise = u'⇑'
+    icon_sunset = u'⇓'
+    icon_temperature = u'○'
+    icon_thunderstorm = u'⛈'
+    icon_wind = u'☴'
+    icons = None
+    lang = 'en'
+    location = None
+    offset_gmt = None
+    request_timeout = 10
+    thresholds = THRESHOLDS
+    unit_rain = 'in'
+    unit_snow = 'in'
+    unit_temperature = 'F'
+    unit_wind = 'mph'
+
+    def _get_icons(self):
+        if self.icons is None:
+            self.icons = {}
+
+        # Defaults for weather ranges
+        defaults = {
+            '200_299': self.icon_thunderstorm,
+            '300_399': self.icon_rain,
+            '500_599': self.icon_rain,
+            '600_699': self.icon_snow,
+            '700_799': self.icon_atmosphere,
+            '800': self.icon_sun,
+            '801_809': self.icon_cloud,
+            '900_909': self.icon_extreme,
+            '950_959': self.icon_wind,
+            '960_999': self.icon_extreme,
+        }
+
+        # Handling ranges from OpenWeatherMap
+        data = {}
+        for source in (defaults, self.icons):
+            for key in source:
+                if not re.match(r'[0-9]+', key):
```

Comment:
(Opinion) We `import re` for one-time use here (only in `post_config_hook`). I think we could avoid `re` by using other options such as...
* if `key.is_digit()` # is `600` only.
* For `600_700`, something like.. if `len(key.split('_')) == 2 and key.replace('_', '').isdigit():`